File: android/app/src/main/java/com/budgetmanager/domain/interactor/PocketInteractor.kt
package com.budgetmanager.domain.interactor

import com.budgetmanager.domain.entity.DomainError
import com.budgetmanager.domain.entity.Pocket
import com.budgetmanager.domain.repository.BudgetRepository
import com.budgetmanager.domain.repository.PocketRepository

class PocketInteractor(
    private val pocketRepository: PocketRepository,
    private val budgetRepository: BudgetRepository
) {
    suspend fun getPockets(): List<Pocket> {
        return pocketRepository.getAll()
    }

    suspend fun getPocket(id: Long): Pocket {
        return pocketRepository.getById(id)
    }

    suspend fun createPocket(
        name: String,
        description: String = "",
        balance: Double = 0.0
    ): Pocket {
        if (name.isBlank()) {
            throw DomainError.InvalidInput("Name cannot be empty")
        }

        val pocket = Pocket(
            name = name,
            description = description,
            balance = balance
        )

        return pocketRepository.add(pocket)
    }

    suspend fun updatePocket(
        id: Long,
        name: String? = null,
        description: String? = null,
        balance: Double? = null
    ): Pocket {
        val pocket = pocketRepository.getById(id)

        val updatedPocket = pocket.copy(
            name = name ?: pocket.name,
            description = description ?: pocket.description,
            balance = balance ?: pocket.balance
        )

        pocketRepository.update(updatedPocket)
        return updatedPocket
    }

    suspend fun deletePocket(id: Long) {
        if (budgetRepository.hasBudgets(id)) {
            throw DomainError.PocketHasBudgets
        }
        pocketRepository.delete(id)
    }

    suspend fun addFunds(id: Long, amount: Double): Pocket {
        if (amount <= 0) {
            throw DomainError.InvalidInput("Amount must be greater than 0")
        }
        pocketRepository.updateBalance(id, amount)
        return pocketRepository.getById(id)
    }
}

File: iOS/Sources/Domain/Interactors/PocketInteractor.swift
import Foundation

class PocketInteractor {
    private let pocketRepository: PocketRepositoryProtocol
    private let budgetRepository: InMemoryBudgetRepository

    init(pocketRepository: PocketRepositoryProtocol, budgetRepository: InMemoryBudgetRepository) {
        self.pocketRepository = pocketRepository
        self.budgetRepository = budgetRepository
    }

    func getPockets() async throws -> [Pocket] {
        return try await pocketRepository.fetchAll()
    }

    func getPocket(id: Int64) async throws -> Pocket {
        return try await pocketRepository.fetch(id: id)
    }

    func createPocket(name: String, description: String = "", balance: Double = 0) async throws -> Pocket {
        guard !name.isEmpty else {
            throw DomainError.invalidInput("Name cannot be empty")
        }

        let pocket = Pocket(name: name, description: description, balance: balance)
        return try await pocketRepository.add(pocket)
    }

    func updatePocket(id: Int64, name: String?, description: String?, balance: Double?) async throws -> Pocket {
        var pocket = try await pocketRepository.fetch(id: id)

        if let name = name {
            pocket.name = name
        }
        if let description = description {
            pocket.description = description
        }
        if let balance = balance {
            pocket.balance = balance
        }

        try await pocketRepository.update(pocket)
        return pocket
    }

    func deletePocket(id: Int64) async throws {
        if budgetRepository.hasBudgets(forPocketId: id) {
            throw DomainError.pocketHasBudgets
        }
        try await pocketRepository.delete(id: id)
    }

    func addFunds(id: Int64, amount: Double) async throws -> Pocket {
        guard amount > 0 else {
            throw DomainError.invalidInput("Amount must be greater than 0")
        }
        try await pocketRepository.updateBalance(id: id, amount: amount)
        return try await pocketRepository.fetch(id: id)
    }
}
